File: Ahorn/entities/AnotherVesselCorruptedBoss.jl
module AnotherVesselCorruptedBoss

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/AnotherVesselCorruptedBoss" AnotherVesselCorruptedBoss(x::Integer, y::Integer, health::Integer=1000, maxHealth::Integer=1000)

const placements = Ahorn.PlacementDict(
    "another_vessel_corrupted_boss" => Ahorn.EntityPlacement(AnotherVesselCorruptedBoss)
)

function Ahorn.selection(entity::AnotherVesselCorruptedBoss)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::AnotherVesselCorruptedBoss, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/another_vessel_corrupted_boss.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local anotherVesselCorruptedBoss = {}

anotherVesselCorruptedBoss.name = "MaggyHelper/AnotherVesselCorruptedBoss"
anotherVesselCorruptedBoss.depth = -8500
anotherVesselCorruptedBoss.justification = {0.5, 1.0}

anotherVesselCorruptedBoss.placements = {
    {
        name = "another_vessel_corrupted_boss",
        data = {
            health = 1000,
            maxHealth = 1000,
    }
    }
}

anotherVesselCorruptedBoss.fieldInformation = {
    health = {
        fieldType = "integer",
        minimumValue = 1
    },
    maxHealth = {
        fieldType = "integer",
        minimumValue = 1
    }
}

anotherVesselCorruptedBoss.fieldOrder = {
    "x", "y",
    "health",
    "maxHealth"
}

function anotherVesselCorruptedBoss.sprite(room, entity)
    local textures = {
        "characters/anothervesselcorrupted/idle00",
        "bgs/maggy/00/anotherhuman/VESSEL",
        "characters/Kglobal::Player/sitDown00"
    }

    for _, texture in ipairs(textures) do
        local ok, sprite = pcall(drawableSprite.fromTexture, texture, entity)
        if ok and sprite then
            sprite:setJustification(0.5, 1.0)
            return {sprite}
        end
    end

    return {}
end

function anotherVesselCorruptedBoss.selection(room, entity)
    return utils.rectangle(entity.x - 18, entity.y - 58, 36, 58)
end

return anotherVesselCorruptedBoss
